docs/spec_draft.md

```diff
@@ -326,6 +327,84 @@ logical operation.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.broadcast_in_dim
+
+### Semantics
+
+Expands the dimensions and/or rank of an input tensor by duplicating the data
+in the `operand` tensor and produces a `result` tensor. Formally, 
+`result[i0, i1, ..., iR-1]` $=$ `operand[j0, j1, ..., jR-1]` such that
+`jk` $=$ `dim(operand, k) == 1 ? 0 : i[boradcast_dimensions[k]]`, for all
+dimensions `k` in `operand`.
+  
+### Operands
+
+| Name                   | Type                                  |
+|------------------------|---------------------------------------|
+| `operand`              | tensor of any supported element types |
+| `broadcast_dimensions` | 1-dimensional array of type `si64`    |
+
+### Results
+
+| Name     | Type                                  |
+|----------|---------------------------------------|
+| `result` | tensor of any supported element types |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the element type.
+  * (C2) dim(`broadcast_dimensions`, 0) $=$ rank(`operand`).
+  * (C3) $0 \le$ `broadcast_dimensions`[i] $\lt$ rank(`result`) for all
+  dimensions i in `operand`.
+  * (C4) All dimensions in `broadcast_dimensions` are unique.
+  * (C5) For all dimensions `i` in `operand`:
+    * `dim(operand, i) = 1` or
+    * `dim(operand, i) = dim(result, j)` where `j = broadcast_dimensions[i]`.  
+
+### Examples
+
+```mlir
+// %operand: [
+//            [
+//             [1, 2],
+//             [3, 4],
+//             [5, 6]
+//            ],
+//            [
+//             [7, 8],
+//             [9, 10],
+//             [11, 12]
+//            ],
+//           ]
+%result = "stablehlo.broadcast_in_dim"(%operand) {
+  broadcast_dimensions = dense<[2, 3, 0]>: tensor<2xi64>
+ } : (tensor<2x3x2xi32>) -> tensor<2x2x2x3xi32>
+// %result: [
+//           [
+//            [
+//             [1, 3, 5],
+//             [7, 9, 11]
+//            ],
+//            [
+//             [1, 3, 5],
+//             [7, 9, 11]
+//            ],
+//           ],
+//           [
+//            [
+//             [2, 4, 6],
+//             [8, 10, 12]
+//            ],
+//            [
+//             [2, 4, 6],
+//             [8, 10, 12]
+//            ],
+//           ]
+//          ]
```

Comment:
What do you think about simplifying the example? This pull request took a lot of effort to figure out, but the readers don't have to know about that :) 

E.g. the proposed example below demonstrates broadcasting, as well as the two special cases: 1) non-sorted dimensions, 2) degenerate broadcasting, while involving fewer distinct numbers to track and fewer dimensions.

```mlir
// %operand: [
//            [1, 2, 3],
//           ]
%result = "stablehlo.broadcast_in_dim"(%operand) {
  broadcast_dimensions = dense<[2, 1]>: tensor<2xi64>
} : (tensor<1x3xi32>) -> tensor<2x3x2xi32>
// %result: [
//            [
//             [1, 1],
//             [2, 2],
//             [3, 3]
//            ],
//            [
//             [1, 1],
//             [2, 2],
//             [3, 3]
//            ],
//          ]
```